crates/bevy_reflect/src/type_registry.rs

```diff
@@ -351,11 +364,33 @@ impl TypeRegistration {
         self.type_info
     }
 
-    /// Inserts an instance of `T` into this registration's type data.
+    /// Inserts an instance of `D` into this registration's type data.
     ///
-    /// If another instance of `T` was previously inserted, it is replaced.
-    pub fn insert<T: TypeData>(&mut self, data: T) {
-        self.data.insert(TypeId::of::<T>(), Box::new(data));
+    /// If another instance of `D` was previously inserted, it is replaced.
+    pub fn register<T: 'static, D: TypeData<T>>(self) -> Self {
```

Comment:
i dislike these signature changes quite a lot. i would suggest instead we make the signature `fn<T: TypeData>(&mut self, data: T) -> &mut Self` so that chaining operations still works. this is what `bevy_app::App` does as well.

Reply:
By this, do you mean that we should revert the `self` change but keep the return `Self` change (as a `&mut Self` instead)?

Reply:
yep